hugr-core/src/export.rs

```diff
@@ -0,0 +1,726 @@
+pub fn export_hugr<'a>(hugr: &'a Hugr, bump: &'a Bump) -> model::Module<'a> {
+    let mut ctx = Context::new(hugr, bump);
+    ctx.export_root();
+    ctx.module
+}
+
+/// State for converting a HUGR graph to its representation in the model.
+struct Context<'a> {
+    /// The HUGR graph to convert.
+    hugr: &'a Hugr,
+    /// The module that is being built.
+    module: model::Module<'a>,
+    /// Mapping from ports to link indices.
+    /// This only includes the minimum port among groups of linked ports.
+    links: IndexSet<(Node, Port)>,
+    bump: &'a Bump,
+}
+
+impl<'a> Context<'a> {
+    pub fn new(hugr: &'a Hugr, bump: &'a Bump) -> Self {
+        let mut module = model::Module::default();
+        module.nodes.reserve(hugr.node_count());
+
+        Self {
+            hugr,
+            module,
+            bump,
+            links: IndexSet::new(),
+        }
+    }
+
+    pub fn export_root(&mut self) {
+        let r#type = self.module.insert_term(model::Term::Wildcard);
+
+        let hugr_children = self.hugr.children(self.hugr.root());
+        let mut children = BumpVec::with_capacity_in(hugr_children.len(), self.bump);
+
+        for child in self.hugr.children(self.hugr.root()) {
+            children.push(self.export_node(child));
+        }
+
+        let root = self.module.insert_region(model::Region {
+            kind: model::RegionKind::DataFlow,
+            sources: &[],
+            targets: &[],
+            children: children.into_bump_slice(),
+            meta: &[],
+            signature: r#type,
+        });
+
+        self.module.root = root;
+    }
+
+    /// Returns the edge id for a given port, creating a new edge if necessary.
+    ///
+    /// Any two ports that are linked will be represented by the same link.
+    fn get_link_id(&mut self, node: Node, port: Port) -> model::LinkId {
+        // To ensure that linked ports are represented by the same edge, we take the minimum port
+        // among all the linked ports, including the one we started with.
+        let linked_ports = self.hugr.linked_ports(node, port);
+        let all_ports = std::iter::once((node, port)).chain(linked_ports);
+        let repr = all_ports.min().unwrap();
+        let edge = self.links.insert_full(repr).0 as _;
+        model::LinkId(edge)
+    }
+
+    pub fn make_ports(&mut self, node: Node, direction: Direction) -> &'a [model::Port<'a>] {
+        let ports = self.hugr.node_ports(node, direction);
+        let mut model_ports = BumpVec::with_capacity_in(ports.len(), self.bump);
+
+        for port in ports {
+            if let Some(model_port) = self.make_port(node, port) {
+                model_ports.push(model_port);
+            }
+        }
+
+        model_ports.into_bump_slice()
+    }
+
+    pub fn make_port(&mut self, node: Node, port: impl Into<Port>) -> Option<model::Port<'a>> {
+        let port: Port = port.into();
+        let op_type = self.hugr.get_optype(node);
```

Comment:
nit: This can be retrieved once in `make_ports` and re-used

Reply:
`make_port` is used once individually as well. Reusing it would just avoid some bounds checks, `get_optype` luckily does not copy everything.